### ISSUE TITLE
Sample Validator contarcts not all functional

### ISSUE THREAD

**Issue by kevinaquick** [paritytech/contracts#96]
We have been setting up a new network of PoA nodes and configuring the spec file to use a Validator Contract. We used one of the simpler sample contracts on the url below as a template for our own:

https://github.com/paritytech/contracts/tree/master/validator_contracts

However when we fired up the nodes Parity kept throwing up an error - on tracking back through the documentation for Validator contracts and the requirements detailed in the ABI we realised we hadn't implemented  the InitiateChange event nor finalizeChange function. SO, this is just to alert you to the fact that not all the contracts you list are current/functional examples and so can be misleading

**Comment by 5chdn** [comment]
ref https://github.com/paritytech/parity/issues/7117

**Comment by rstormsf** [comment]
@kevinaquick were you able to deploy and verify the contracts? 
were you able to vote for a validator because their `addSupport` function immediately tries to call `addValidator` which is not the way I'd think it should work

**Comment by kevinaquick** [comment]
Hi Roman, we created our own version of the validator contract based on our current needs, so the addSupport doesn't exist in ours nor innerset or outerset - we did eventually get ours all working as we wished, once we had fully complied with the ABI requirements and  set the flag in the spec file to "safeContract" rather than "contract" as ours is a non-logging example (it tooks a while to figure out what the documentation was saying especially when compared to some of the example contracts).   

In our private network at the moment we didn't want the vote based system for adding validators (we may eventually need to change that as the network spreads, but it was a complexity we didn't need for now). Our contract basically has admin accounts which are permitted to add new validators with functions to add/remove validators to add/remove administrators and controls on minimum numbers of admins to reduce the chance the contract getting into a state of having no admins or operational admin accounts

**Comment by rstormsf** [comment]
@kevinaquick why don't you share your contract?

**Comment by msbachler** [comment]
Our contract its just a simple validator list with admins (posted on behalf of @kevinaquick).

pragma solidity ^0.4.11;

contract AdminValidatorList {
    	address public owner;

	address[] public validatorList = [<your first validator address>];
	address[] public pendingList = [<your first validator address>];

	mapping(address => bool) public isAdmin;

	modifier onlyAdmin() {
		require(isAdmin[msg.sender] == true);
		_;
	}
	
	event validatorAdded(address newvalidator);
	event validatorRemoved(address oldvalidator);
	event adminAdded(address newadmin);
	event adminRemoved(address oldadmin);
	event InitiateChange(bytes32 indexed _parent_hash, address[] _new_set);
	
	function AdminValidatorList() public {
		isAdmin[validatorList[0]] = true;		
	}

	// Called on every block to update node validator list.
	function getValidators() public constant returns (address[] _validators) {
		return validatorList;
	}

	function getPendingValidators() public constant returns (address[] _p) {
		return pendingList;
	}

	// Add a validator to the list.
	function addValidator(address validator) public onlyAdmin {
		for (uint i = 0; i < pendingList.length; i++) {
			require(pendingList[i] != validator);
		}

		pendingList.push(validator);
		validatorAdded(validator);
		InitiateChange(block.blockhash(block.number - 1),pendingList);
	}

	// Remove a validator from the list.
	function removeValidator(address validator) public onlyAdmin returns (bool success){
				
		uint i=0;
		uint count = pendingList.length;
		
		// you don't want to leave no validators - can't delete any until you have a minimum of 3. 
		// This is in case your 1 remaining node goes down. Leave a safety margin of 2
		if (count > 2) { 
			success = false;
			for (i=0; i<count;i++) {
				if (pendingList[i] == validator) {
					if (i < pendingList.length-1) {
						pendingList[i] = pendingList[pendingList.length-1];
					}
					pendingList.length--;
					success = true;
					validatorRemoved(validator);
					InitiateChange(block.blockhash(block.number - 1),pendingList);
					break;
				}
			}
		}
		return success;
	}

	// Add an admin.
	function addAdmin(address admin) public onlyAdmin {
		isAdmin[admin] = true;
		adminAdded(admin);
	}

	// Remove an admin.
	function removeAdmin(address admin) public onlyAdmin {
		isAdmin[admin] = false;
		adminRemoved(admin);
	}
	
	function finalizeChange() public {
		validatorList = pendingList;
	}
}

**Comment by rphmeier** [comment]
Yes, it's true that some of these contracts are not updated to the recent ABI updates for the validator set. We plan to upgrade them at some point but for now we're focusing on simpler, more easily audited contracts.

I wouldn't describe these as samples as much as in-development contracts that we actually intend to use. We can provide some small example sets which will be more comprehensive.

@msbachler 

Note that the `finalizeChange` function should check that `msg.sender` is the `SYSTEM_ADDRESS` that no-one actually holds the key for. It is meant only to be called by the state transition system (during the same phase as block rewards are applied, making invisible changes to the state without corresponding transaction). Otherwise anyone can call `finalizeChange`.

**Comment by msbachler** [comment]
Thank you. I will change my contract. I see the example contracts are now updated. That helps enormously.

**Comment by msbachler** [comment]
When I compile my newly edited contract I get:

```
Warning: This looks like an address but has an invalid checksum. 
If this is not used as an address, please prepend '00'.
        address constant SYSTEM_ADDRESS = 0xfffffffffffffffffffffffffffffffffffffffe;
```
I ran an npm install solc again to make sure I was on the latest version, 0.4.19, but I get the same error. This line of code is used in the example contracts  RelaySet, OwnedSet and MajoritySet which say they are 0.4.15. The last time I had this error it was because I was using an address it didn't recognize (from a different chain - by mistake). But I thought that this was a system set address. Does it need to be added to the spec.json of the nodes on the chain or something to make it recognized or is it a different issue?

**Comment by msbachler** [comment]
Never mined. I just remembered I have to run all address I put into contracts through my checksum function first. It now compiles fine as:
`address constant SYSTEM_ADDRESS = 0xffffFFFfFFffffffffffffffFfFFFfffFFFfFFfE;`
Sorry about that. I just forgot about that step.

**Comment by msbachler** [comment]
Might be worth putting a comment in those three contract above the SYSTEM_ADDRESS to explain that to people.

**Comment by msbachler** [comment]
In case anyone is interested this is the part of the Solidity compiler that checks addresses in your contracts: https://github.com/ethereum/solidity/blob/2b5ef80634c94a5cf27f3ce8a85eeb9d713f5ec3/libdevcore/CommonData.cpp

Look specifically at the functions 'passesAddressChecksum' and 'getChecksummedAddress'
